src/server/pegasus_server_impl.cpp

```diff
@@ -670,41 +673,44 @@ void pegasus_server_impl::on_multi_get(multi_get_rpc rpc)
                                      rpc.remote_address(),
                                      status.ToString());
                 }
-            }
-            // check ttl
-            if (status.ok()) {
-                uint32_t expire_ts = pegasus_extract_expire_ts(_pegasus_data_version, value);
-                if (expire_ts > 0 && expire_ts <= epoch_now) {
-                    expire_count++;
-                    if (_verbose_log) {
-                        LOG_ERROR_PREFIX("rocksdb data expired for multi_get from {}",
-                                         rpc.remote_address());
-                    }
-                    status = rocksdb::Status::NotFound();
+
+                if (status.IsNotFound()) {
+                    continue;
                 }
+
+                error_occurred = true;
+                final_status = status;
+                break;
             }
-            // extract value
-            if (status.ok()) {
-                // check if exceed limit
-                if (count >= max_kv_count || size >= max_kv_size) {
-                    exceed_limit = true;
-                    break;
-                }
-                ::dsn::apps::key_value kv;
-                kv.key = request.sort_keys[i];
-                if (!request.no_value) {
-                    pegasus_extract_user_data(_pegasus_data_version, std::move(value), kv.value);
+
+            // check ttl
+            if (check_if_record_expired(epoch_now, value)) {
+                expire_count++;
+                if (_verbose_log) {
+                    LOG_ERROR_PREFIX(
+                        "rocksdb data expired for multi_get from {}: hash_key = \"{}\", "
+                        "sort_key = \"{}\"",
+                        rpc.remote_address(),
+                        ::pegasus::utils::c_escape_string(request.hash_key),
+                        ::pegasus::utils::c_escape_string(request.sort_keys[i]));
                 }
-                count++;
-                size += kv.key.length() + kv.value.length();
-                resp.kvs.emplace_back(std::move(kv));
+                status = rocksdb::Status::NotFound();
+                continue;
             }
-            // if error occurred
-            if (!status.ok() && !status.IsNotFound()) {
-                error_occurred = true;
-                final_status = status;
+
+            // check if exceed limit
+            if (count >= max_kv_count || size >= max_kv_size) {
```

Comment:
How about add `dsn_likely/unlikey` in hot paths?

Reply:
OK, add `dsn_unlikely` here.